common/oauthTokenManager.go

```diff
@@ -142,6 +149,203 @@ func (uotm *UserOAuthTokenManager) MSILogin(ctx context.Context, identityInfo Id
+	oAuthTokenInfo, err := secretLoginNoUOTM(tenantID, activeDirectoryEndpoint, secret, applicationID)
+
+	if persist {
+		err = uotm.credCache.SaveToken(*oAuthTokenInfo)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return oAuthTokenInfo, nil
+}
+
+//GetNewTokenFromSecret is a refresh shell for secretLoginNoUOTM
+func (credInfo *OAuthTokenInfo) GetNewTokenFromSecret(ctx context.Context) (*adal.Token, error) {
+	tokeninfo, err := secretLoginNoUOTM(credInfo.Tenant, credInfo.ActiveDirectoryEndpoint, credInfo.SPNInfo.Secret, credInfo.ApplicationID)
+
+	if err != nil {
+		return nil, err
+	} else {
+		return &tokeninfo.Token, nil
+	}
+}
+
+func certLoginNoUOTM(tenantID, activeDirectoryEndpoint, certPath, certPass, applicationID string) (*OAuthTokenInfo, error) {
+	if tenantID == "" {
+		tenantID = DefaultTenantID
+	}
+
+	if activeDirectoryEndpoint == "" {
+		activeDirectoryEndpoint = DefaultActiveDirectoryEndpoint
+	}
+
+	if applicationID == "" {
+		return nil, fmt.Errorf("please supply your OWN application ID")
+	}
+
+	oAuthTokenInfo := OAuthTokenInfo{
+		Tenant:                  tenantID,
+		ActiveDirectoryEndpoint: activeDirectoryEndpoint,
+	}
+
+	oauthConfig, err := adal.NewOAuthConfig(activeDirectoryEndpoint, tenantID)
+	if err != nil {
+		return nil, err
+	}
+
+	certData, err := ioutil.ReadFile(certPath)
+	if err != nil {
+		return nil, err
+	}
+
+	var pk interface{}
+	var cert *x509.Certificate
+
+	if path.Ext(certPath) == ".pfx" {
+		pk, cert, err = pkcs12.Decode(certData, certPass)
+		if err != nil {
+			return nil, err
+		}
+	} else if path.Ext(certPath) == ".pem" {
+		block, rest := pem.Decode(certData)
+
+		for len(rest) != 0 || pk == nil || cert == nil {
+			switch block.Type {
+			case "PRIVATE KEY":
+				pk, err = x509.ParsePKCS8PrivateKey(block.Bytes)
+
+				if err != nil {
+					return nil, fmt.Errorf("private key has invalid format")
+				}
+			case "CERTIFICATE":
+				cert, err = x509.ParseCertificate(block.Bytes)
+
+				if err != nil {
+					return nil, fmt.Errorf("certificate has invalid format")
+				}
+			default:
+				return nil, fmt.Errorf("field " + block.Type + " not expected")
+			}
+
+			if len(rest) == 0 && (pk == nil || cert == nil) {
+				return nil, fmt.Errorf("could not find the required information (private key & certificate) in the supplied .pem file")
+			}
+
+			block, rest = pem.Decode(rest)
+		}
+	} else {
+		return nil, fmt.Errorf("please supply either a .pfx or a .pem file containing a private key and a certificate")
+	}
+
+	spt, err := adal.NewServicePrincipalTokenFromCertificate(
+		*oauthConfig,
+		applicationID,
+		cert,
+		pk.(*rsa.PrivateKey),
+		Resource,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	err = spt.Refresh()
+	if err != nil {
+		return nil, err
+	}
+
+	cpfq, _ := filepath.Abs(certPath)
+
+	oAuthTokenInfo.Token = spt.Token()
+	oAuthTokenInfo.RefreshToken = oAuthTokenInfo.Token.RefreshToken
```

Comment:
Wouldn't this be the same as the secret-based auth, above, in that we don't expect there to be any refersh token? If so, maybe for the sake of clarity align this with the secret-based code and replace this line with a comment descrbing that we do not expect to recieve any refresh token from the Service.